### PR TITLE
Feat/samples wasm

### PR DESCRIPTION
**Pull request by Odonno**

### CODE REVIEW

**Comment by Roms1383** [comment]
Sorry I can't help reviewing here because I haven't done any Svelte yet, and not much WASM ^^

As a side note what I had in mind initially is a bit different: use `surrealdb-migrations` on e.g. Android, iOS, ... with `flutter_rust_bridge` crate. Mostly like one would with SQLite.

Essentially the migrations steps are almost the same everywhere: that'd be cool to provide it for all (even gradually): `ws` / `wss` / `http` / `inflxdb` / `rocksdb` / etc.

When I tried with `rocksdb` locally there's minor differences like auth and server are not available as far as I understood.

File: samples/wasm/rsw-hello/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        .use_db("test")
+        .await
+        .expect("Failed to use namespace");
+
+    SurrealdbMigrations::new(&db)
+        .load_files(&DB_DIR)
+        .up()
+        .await
+        .expect("Failed to apply migrations");
+}
```
Comment (Roms1383) [inline samples/wasm/rsw-hello/src/lib.rs:39]: That kind of ergonomic is exactly what I'm looking for :) Looks promising

**Comment by Odonno** [comment]
> Sorry I can't help reviewing here because I haven't done any Svelte yet, and not much WASM ^^

No worry, this is a basic sample app.

> As a side note what I had in mind initially is a bit different: use `surrealdb-migrations` on e.g. Android, iOS, ... with `flutter_rust_bridge` crate. Mostly like one would with SQLite.

I see. I do not have any experience with Flutter so it will be hard for me to provide any guidance on this but I suppose it should follow the same pattern as WASM. However, I set the function `can_use_filesystem` conditioned by the target_arch, it will return `false` on WASM and `true` on every other. I do not know what target arch is used with your `flutter_rust_bridge` crate. If they do not share the same target, it is highly possible that the lib will fail.

> Essentially the migrations steps are almost the same everywhere: that'd be cool to provide it for all (even gradually): `ws` / `wss` / `http` / `inflxdb` / `rocksdb` / etc.

Indeed, the pattern should be essentially the same whether you use X or Y embedded solution. If you want to, you could provide your own samples once this PR is merged.

> When I tried with `rocksdb` locally there's minor differences like auth and server are not available as far as I understood.

Indeed, when you use a local solution like IndexedDb, auth is not available because you have direct access to the embedded file. That makes sense because this is a totally useless feature in this context. However, I do not know what you meant by `server`?

**Comment by Roms1383** [comment]
> I see. I do not have any experience with Flutter so it will be hard for me to provide any guidance on this but I suppose it should follow the same pattern as WASM. However, I set the function can_use_filesystem conditioned by the target_arch, it will return false on WASM and true on every other. I do not know what target arch is used with your flutter_rust_bridge crate. If they do not share the same target, it is highly possible that the lib will fail.

There's nothing much really, it's compiled to Android, iOS, Windows, Linux, etc and `flutter_rust_bridge` auto-generate safe FFI wrapper to allow bridging between Flutter (Dart) and Rust. I agree it should be very similar to any other native target (like when you build `surrealdb-migrations` on `ubuntu` let's say). I haven't checked `can_use_filesystem` yet but I guess by its name that it's more dependent on the permissions requested by the app in Android, iOS, etc for it to succeed.

I already did try it without migrations and it works as far as I can tell (`connect` to local rocksdb, then `query` a bunch of `DEFINE`s then `select`).

> Indeed, the pattern should be essentially the same whether you use X or Y embedded solution. If you want to, you could provide your own samples once this PR is merged.

Can't promise, but yes :)

> Indeed, when you use a local solution like IndexedDb, auth is not available because you have direct access to the embedded file. That makes sense because this is a totally useless feature in this context.

Utterly.

> However, I do not know what you meant by server?

Ah yes I meant that you run e.g. `surreal start ...` while I don't think it's needed to open e.g. a `rocksdb` file and `.query` some migrations on it. Or did I miss something ?

**Comment by Roms1383** [comment]
Ah I should add something: basically when you build for native targets in Flutter, as far as Rust is concerned it's like building normally. And when Flutter targets the Web, then Rust also has to be build with WASM.

**Comment by Odonno** [comment]
OK, so I have to find a way to figure out when accessing the filesystem is a supported feature or not. I mean, it is clearly unsupported on wasm but what about the other platforms flutter support?

**Comment by Roms1383** [comment]
My assumption is as long as we can do like this https://github.com/Odonno/surrealdb-migrations/pull/23#discussion_r1199599205 it should just works if the permissions for the platform have been correctly set (e.g. Android permissions to [access filesystem](https://developer.android.com/training/data-storage/manage-all-files?hl=en), [shared storage](https://developer.android.com/training/data-storage/shared/documents-files?hl=en), etc)

**Comment by Roms1383** [comment]
Looking at your code I don't think you need to change anything actually.

**Comment by Odonno** [comment]
I think I found a better way to know if I have access to the filesystem. It should be working fine on every context now.

**Pull request merged**